### DISCUSSION TITLE
Why is this happening? "stateEstimate"

### DISCUSSION THREAD

**Discussion by GNC-JY** [bitcraze/discussions#630] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/110527007/227203388-ea5aefbf-1af0-421a-aaac-1f5533fffbdf.png)

Hello.

Like the picture above, in the stateEstimate.roll and stateEstimate.pitch variables, the pitch was raised, and it was confirmed that it was initialized at around 90 degrees. Curious why it doesn't go beyond 90 degrees, and where is the location of the code related to this?

Is it supposed to be reset to zero when it reaches 90 degrees to prevent gimbal lock?

And there were cases where the pitch was moved, but the roll was also changed or the value was instantaneously diverged.

**Comment by cc299792458** [comment, marked as answer]
Do you use kalman estimator or complementary estimator? There is a reset in kalman estimator if it exceed the bounds. 
But in your case, it resets regularly so I think it not seems like kalman estimator reset.

**Reply by GNC-JY** [reply]
I used to use flow deck, but after removing flow deck, the stateEstimate value comes out normally. What I'm curious about here is that pitch is -90 to 90 degrees, but why is roll -180 to 180 degrees?

**Reply by GNC-JY** [reply]
ah, I found it.

![image](https://user-images.githubusercontent.com/110527007/227466995-61370810-c7b7-4ab5-87e9-bde75a040bb2.png)
asin range is -90 to 90.

**Reply by krichardsson** [reply]
I'm happy you found your answer!

For general information: the kalman estimator is automatically enabled when the flow deck (and other positioning decks) is used. You'll find the line `ESTIMATOR: Using Kalman (2) estimator` in the console log with the flow deck, and `ESTIMATOR: Using Complementary (1) estimator` without it.

**Reply by GNC-JY** [reply]
So, why is estimatePitch resetting at 90 degrees like in the first picture? I think it's to prevent the diverges of atan and asin, Where is the location of the code?

**Reply by krichardsson** [reply]
The 90 degrees you see is most likely a coincidence. The kalman filter resets if the position is > 100 m or velocity > 10 m/s, if you plot position and velocity you will most likely see that this is the case. The reset is triggered by a call to [`kalmanSupervisorIsStateWithinBounds()`](https://github.com/bitcraze/crazyflie-firmware/blob/889ff485908ad80627fd73158ee8449abf3e35bb/src/modules/src/kalman_supervisor.c#L37-L57)

By the way, `sensfusion6GetEulerPRY()` is only used in the complementary filter, not the Kalman estimator

**Reply by GNC-JY** [reply]
yes. I'm already found it in kalman_core.c and change this one.
![image](https://user-images.githubusercontent.com/110527007/228158348-1c243252-eb43-4a31-a15a-68d15c4d609c.png)

**Reply by GNC-JY** [reply]
Other experiments have shown that if I stay in a twisted attitude for a long time, the value seems to be reset(Initialized). Where is the code for this part in the firmware?

**Reply by krichardsson** [reply]
The only code that triggers a reset is the [`kalmanSupervisorIsStateWithinBounds()`](https://github.com/bitcraze/crazyflie-firmware/blob/889ff485908ad80627fd73158ee8449abf3e35bb/src/modules/src/kalman_supervisor.c#L37-L57) as far as I know. 

There is a feature that "leaks" the yaw towards zero, but this is a slow process and not a hard reset. This feature is important for positioning decks that can not track absolute yaw, like the Loco deck.

**Reply by GNC-JY** [reply]
Okay. I got it. Thank you very much.
